src/main/java/com/nowoczesnyjunior/financialapp/expensemodule/service/ExpenseServiceImpl.java

```diff
@@ -0,0 +1,111 @@
+import com.nowoczesnyjunior.financialapp.expensemodule.mapper.ExpenseMapper;
+import com.nowoczesnyjunior.financialapp.expensemodule.model.Expense;
+import com.nowoczesnyjunior.financialapp.expensemodule.repository.CategoryRepository;
+import com.nowoczesnyjunior.financialapp.expensemodule.repository.ExpenseRepository;
+import com.nowoczesnyjunior.financialapp.openapi.model.CategoryDto;
+import com.nowoczesnyjunior.financialapp.openapi.model.ExpenseDto;
+import com.nowoczesnyjunior.financialapp.usermodule.model.User;
+import com.nowoczesnyjunior.financialapp.usermodule.repository.UserRepository;
+import org.hibernate.ObjectNotFoundException;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class ExpenseServiceImpl implements ExpenseService {
+
+    private final ExpenseRepository expenseRepository;
+    private final CategoryRepository categoryRepository;
+    private final UserRepository userRepository;
+    private final ExpenseMapper expenseMapper;
+
+    LocalDateTime defaultStartDate = LocalDateTime.of(1970, 01, 01, 00, 00);
+    LocalDateTime defaultEndDate = LocalDate.now().atStartOfDay();
+
+    public ExpenseServiceImpl(ExpenseRepository expenseRepository, CategoryRepository categoryRepository, UserRepository userRepository, ExpenseMapper expenseMapper) {
+        this.expenseRepository = expenseRepository;
+        this.categoryRepository = categoryRepository;
+        this.userRepository = userRepository;
+        this.expenseMapper = expenseMapper;
+    }
+
+    @Override
+    public List<ExpenseDto> getExpenses(String startDate, String endDate, String categoryName, Integer topN) {
+        LocalDateTime startLocalDate = startDate != null ? getLocalDateTime(startDate) : defaultStartDate;
+        LocalDateTime endLocalDate = endDate != null ? getLocalDateTime(endDate) : defaultEndDate;
+        List<Expense> list;
+
+        if (categoryName != null) {
+            validateCategory(categoryName);
+            list = expenseRepository.findExpenseByExpenseDateBetweenAndCategory(startLocalDate, endLocalDate, categoryName);
+        } else {
+            list = expenseRepository.findExpenseByExpenseDateBetween(startLocalDate, endLocalDate);
+        }
+        return expenseMapper.expensesToDtos(list);
+    }
+
+    @Override
+    public ExpenseDto addExpense(ExpenseDto expenseDto) {
+        Expense expense = expenseMapper.dtoToModel(expenseDto);
+        // TODO: get active user from usermodule
+        User user = userRepository.findAll().get(0);
+        expense.setUser(user);
+
+        if (expenseDto.getCategory().getId() == null || !isCategoryAvailable(expenseDto.getCategory())) {
+            throw new CategoryNotFoundException();
+        }
+
+        Expense savedExpense = expenseRepository.save(expense);
+        return expenseMapper.expenseToExpenseDto(savedExpense);
+    }
```

Comment:
Here I would move the mapping below if statement, if category doesn't exist, we don't need to perform this mapping
Also, maybe we should deal with all validation in separate method? There would be probably more?

Reply:
Here I am not really sure, what you mean:
- in "if" statement I am checking if the category exists
- if it is not then the exception is thrown and the following code is not excetued (also the mapping)

or have I missed something?

Reply:
What I meant is that this entire block of code before 'if' (with expense mapping and user search) is not used if exception is thrown, so it might be moved below, like this:

 ```
@Override
    public ExpenseDto addExpense(ExpenseDto expenseDto) {
        if (expenseDto.getCategory().getId() == null || !categoryExists(expenseDto.getCategory())) {
            throw new CategoryNotFoundException();
        }

        Expense expense = expenseMapper.toModel(expenseDto);
        // TODO: get active user from usermodule
        User user = userRepository.findAll().get(0);
        expense.setUser(user);
        
        Expense savedExpense = expenseRepository.save(expense);
        return expenseMapper.toDto(savedExpense);
    }
```